src/tactic/polyrith.lean

```diff
@@ -0,0 +1,605 @@
+
+/--
+A parser object that parses `string`s of the form `"poly.mul p q"`
+to the appropriate `poly` object representing the product of two `poly`s.
+Here, `p` and `q` are themselves string forms of `poly`s.
+-/
+meta def mul_parser (cont : parser poly) : parser poly := do
+  str "poly.mul ",
+  lhs ← cont,
+  ch ' ',
+  rhs ← cont,
+  return (poly.mul lhs rhs)
+
+/--
+A parser object that parses `string`s of the form `"poly.pow p n"`
+to the appropriate `poly` object representing a `poly` raised to the
+power of a natural number. Here, `p` is the string form of a `poly`
+and `n` is a natural number.
+-/
+meta def pow_parser (cont : parser poly) : parser poly := do
+  str "poly.pow ",
+  base ← cont,
+  ch ' ',
+  n ← nat,
+  return (poly.pow base n)
+
+/--
+A parser object that parses `string`s into `poly` objects.
+-/
+meta def poly_parser : parser poly := do
+  ch '(',
+  t ←  var_parser <|> const_fraction_parser <|> add_parser poly_parser
+    <|> sub_parser poly_parser <|> mul_parser poly_parser <|> pow_parser poly_parser,
+  ch ')',
+  return t
+
+/--A parser object that parses a string (which may contain trailing whitespace) into a poly object-/
+meta def one_of_many_poly_parser : parser poly := do
+  p ← poly_parser,
+  optional $ ch ' ',
+  return p
+
+/--checks whether a character is a whitespace character-/
+@[derive decidable]
+meta def _root_.char.is_whitespace' (c : char) : Prop :=
+c.is_whitespace ∨ c.to_nat = 13
+
+/--removes trailing whitespace from a `string`-/
+meta def remove_trailing_whitespace : string → string
+| s := if s.back.is_whitespace' then remove_trailing_whitespace s.pop_back else s
+
+/--
+A parser object that converts the output of sage (which should be
+formatted as a collection of ``poly`s in `string` form, separated
+ by a single space) into a `list poly`
+-/
+meta def sage_output_parser : parser (sage_error ⊕ list poly) :=
+(sum.inl <$> error_parser) <|> sum.inr <$> many (one_of_many_poly_parser)
+
+/--A tactic that checks whether `sage_output_parser` worked-/
+meta def parser_output_checker : string ⊕ (sage_error ⊕ list poly) → tactic (list poly)
+| (sum.inl s) := fail!"polyrith failed to parse the output from Sage.\n\n{s}"
+| (sum.inr (sum.inr poly_list)) := return poly_list
+| (sum.inr (sum.inl err))
+  := fail!"polyrith failed to retrieve a solution from Sage! {err.kind}: {err.message}"
+
+/--
+A tactic that uses the above defined parsers to convert
+the sage output `string` to a `list poly`
+-/
+meta def convert_sage_output : string → tactic (list poly)
+|s := (let sage_stuff := sage_output_parser.run_string (remove_trailing_whitespace s)
+  in parser_output_checker sage_stuff)
+
+
+/-!
+# Parsing context into poly
+
+The following section contains code that collects hypotheses of the appropriate type
+from the context (and from the list of hypotheses and proof terms specified by the user)
+and converts them into `poly` objects.
+-/
+
+local notation `reduc` := transparency.reducible
```

Comment:
This is only used twice, as far as I can tell. I think it would be better not to introduce the notation and just write it out in full.

Reply:
Fixed!